lib/steep/type_construction.rb

```diff
@@ -2731,6 +2731,10 @@ def type_lambda(node, params_node:, body_node:, type_hint:)
         )
 
         if expected_block_type = block_constr.block_context.body_type
+          type_vars = expected_block_type.free_variables
```

Comment:
This is the fix.

The type error is caused by type checking lambda return type with given hint, which was type variable.
The type variable in the hint doesn't make any sense, because the lambda will be type checked with the hint in next step. The change apply a substitution that replaces all type variables to `untyped`.